crates/wit-component/src/printing.rs

```diff
@@ -990,66 +1115,223 @@ fn is_keyword(name: &str) -> bool {
+/// modeled after the support in `wit-bindgen-core`. Indentation is set to two spaces.
 #[derive(Default)]
-struct Output {
+pub struct OutputToString {
     indent: usize,
     output: String,
+    // set to true after newline, then to false after first item is indented.
+    needs_indent: bool,
 }
 
-impl Output {
-    fn push_str(&mut self, src: &str) {
-        let lines = src.lines().collect::<Vec<_>>();
-        for (i, line) in lines.iter().enumerate() {
-            let trimmed = line.trim();
-            if trimmed.starts_with('}') && self.output.ends_with("  ") {
-                self.output.pop();
-                self.output.pop();
-            }
-            self.output.push_str(if lines.len() == 1 {
-                line
-            } else {
-                line.trim_start()
-            });
-            if trimmed.ends_with('{') {
-                self.indent += 1;
-            }
-            if trimmed.starts_with('}') {
-                // Note that a `saturating_sub` is used here to prevent a panic
-                // here in the case of invalid code being generated in debug
-                // mode. It's typically easier to debug those issues through
-                // looking at the source code rather than getting a panic.
-                self.indent = self.indent.saturating_sub(1);
-            }
-            if i != lines.len() - 1 || src.ends_with('\n') {
-                // Trim trailing whitespace, if any, then push an indented
-                // newline
-                while let Some(c) = self.output.chars().next_back() {
-                    if c.is_whitespace() && c != '\n' {
-                        self.output.pop();
-                    } else {
-                        break;
-                    }
-                }
-                self.output.push('\n');
-                for _ in 0..self.indent {
-                    self.output.push_str("  ");
-                }
+impl OutputToString {
+    fn indent_if_needed(&mut self) {
+        if self.needs_indent {
+            for _ in 0..self.indent {
+                // Indenting by two spaces.
+                self.output.push_str("  ");
             }
+            self.needs_indent = false;
+        }
+    }
+
+    fn indent_and_print(&mut self, src: &str) {
+        assert!(!src.contains('\n'));
+        if src.starts_with(' ') {
+            assert!(
+                !self.needs_indent,
+                "cannot add a space at the begining of a line"
+            );
         }
+        self.indent_if_needed();
+        self.output.push_str(src);
     }
 }
 
-impl Write for Output {
-    fn write_str(&mut self, s: &str) -> fmt::Result {
-        self.push_str(s);
-        Ok(())
+impl Output for OutputToString {
+    fn newline(&mut self) {
+        self.output.push('\n');
+        self.needs_indent = true;
+    }
+
+    fn keyword(&mut self, src: &str) {
+        self.indent_and_print(src);
+    }
+
+    fn r#type(&mut self, src: &str, _kind: TypeKind) {
```

Comment:
Could this be named perhaps `ty` or `type_`? (personally I prefer to avoid `r#foo` where possible)

Reply:
Sure, I've renamed it to `ty`.